daffodil-test/src/test/resources/org/apache/daffodil/section23/dfdl_expressions/expressions3.tdml

```diff
@@ -0,0 +1,400 @@
+            <foo>
+              <hdr>2</hdr>
+              <bar>1</bar>
+              <c2>b</c2>
+            </foo>
+          </e2>
+        </ex:r>
+      </tdml:dfdlInfoset>
+    </tdml:infoset>
+  </tdml:parserTestCase>
+  
+  <tdml:defineSchema name="s2b" elementFormDefault="unqualified">
+    <xs:include schemaLocation="org/apache/daffodil/xsd/DFDLGeneralFormat.dfdl.xsd" />
+    <dfdl:format ref="ex:GeneralFormat" lengthKind="delimited"/>
+
+    <xs:group name="g">
+      <!-- 
+        The path expression below is depending on the ability to 
+        compile polymorphically. The expression ../foo/hdr is always an int.
+        The expression ../foo/bar is a string, or an integer depending on 
+        the context. 
+       -->
+      <xs:sequence>
+      <xs:choice dfdl:choiceDispatchKey="{ 
+         xs:string(if (../foo/hdr eq 1) 
+           then xs:int( ../foo/bar )
+           else xs:int( ../foo/bar + 1 ) )}">
+        <xs:element name="c1" type="xs:string" dfdl:choiceBranchKey="1" />
+        <xs:element name="c2" type="xs:string" dfdl:choiceBranchKey="2" />
+      </xs:choice>
+      </xs:sequence>
+    </xs:group>
+
+    <xs:element name="r">
+      <xs:complexType>
+        <xs:sequence dfdl:separator="|">
+          <xs:element name="e1">
+            <xs:complexType>
+              <xs:sequence>
+                <xs:element name="foo">
+                  <xs:complexType>
+                    <xs:sequence dfdl:separator="|">
+                      <xs:element name="hdr" type="xs:int"/>
+                      <xs:element name="bar" type="xs:float" />
+                      <xs:group ref="ex:g" />
+                    </xs:sequence>
+                  </xs:complexType>
+                </xs:element>
+              </xs:sequence>
+            </xs:complexType>
+          </xs:element>
+          <xs:element name="e2">
+            <xs:complexType>
+              <xs:sequence>
+                <xs:element name="foo">
+                  <xs:complexType>
+                    <xs:sequence dfdl:separator="|">
+                      <xs:element name="hdr" type="xs:int"/>
+                      <xs:element name="bar" type="xs:int" />
+                      <xs:group ref="ex:g" />
+                    </xs:sequence>
+                  </xs:complexType>
+                </xs:element>
+              </xs:sequence>
+            </xs:complexType>
+          </xs:element>
+        </xs:sequence>
+      </xs:complexType>
+    </xs:element>
+ </tdml:defineSchema>
+
+  <tdml:parserTestCase name="test_polymorphic_expr_2b" model="s2b" root="r"
+    description="Polymorphic path expression">
+    <tdml:document>1|1|a|2|1|b</tdml:document>
+    <tdml:errors>
+      <tdml:error>Schema Definition Error</tdml:error>
+      <tdml:error>expression ../foo/bar</tdml:error>
+      <tdml:error>SignedNumeric</tdml:error>
+      <tdml:error>cannot be converted to xs:int</tdml:error>
+      <tdml:error>element bar with xs:int type at Location</tdml:error>
```

Comment:
There are actually two diagnostics about different element bar. They have different locations in the TDML schema. The TDML runner can't specify that there are two distinct matches for the string.